[pytorch][counters] Pybind for WaitCounter

andriigrynenko opened the pull request:
Summary:
Basic pybind integration for WaitCounter providing a guard API.
Also fixes broken copy/move constructor in WaitGuard (it wasn't really used with the macro-based C++ API).

Test Plan: unit test

Reviewed By: asiab4

Differential Revision: D60463979

Skylion007 commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -296,6 +298,48 @@ void initMonitorBindings(PyObject* module) {
+      m, "WaitCounterTracker")
+      .def(
+          "__enter__",
+          [](std::shared_ptr<WaitCounterTracker> self) {
```
```suggestion
          [](std::shared_ptr<WaitCounterTracker>& self) {
```

Skylion007 commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -296,6 +298,48 @@ void initMonitorBindings(PyObject* module) {
+          })
+      .def(
+          "__exit__",
+          [](std::shared_ptr<WaitCounterTracker> self, const pybind11::args&) {
```
```suggestion
          [](std::shared_ptr<WaitCounterTracker>& self, const pybind11::args&) {
```

andriigrynenko committed:
[pytorch][counters] Pybind for WaitCounter (#132167)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/132167

Basic pybind integration for WaitCounter providing a guard API.
Also fixes broken copy/move constructor in WaitGuard (it wasn't really used with the macro-based C++ API).

Test Plan: unit test

Reviewed By: asiab4

Differential Revision: D60463979

clee2000 commented:
@pytorchbot revert -m "broke test_public_bindings.py::TestPublicBindings::test_correct_module_names [GH job link](https://github.com/pytorch/pytorch/actions/runs/10183687967/job/28172929836) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/2c7bd61afa4b762e00b26bbde43685de080af32a) not tested on PR due to bad TD" -c weird
